2022Swerve/src/main/java/frc/robot/subsystems/AutonSubsytem.java

```diff
@@ -0,0 +1,184 @@
+        this.swerveSubsystem = swerveSubsystem;
+
+        //Add auton modes to chooser
+        for(AutonModes mode : AutonModes.values()){
+            modeChooser.addOption(mode.toString(), mode);
+        }
+        modeChooser.setDefaultOption(kDefaultAutonMode.toString(), kDefaultAutonMode);
+        configTab.add("Auton mode", modeChooser);
+
+        //Define PID controllers for tracking trajectory
+        xController = new PIDController(AutonConstants.kXTranslationP, 0, 0);
+        yController = new PIDController(AutonConstants.kYTranslationP, 0, 0);
+        rotationController = new PIDController(AutonConstants.kRotationP, 0, 0);
+        rotationController.enableContinuousInput(-Math.PI, Math.PI);
+
+        pathConstraints = new PathConstraints(AutonConstants.kMaxSpeed, AutonConstants.kMaxAcceleration);
+    }
+
+    private PathPlannerTrajectory getTrajectory(String name) throws NullPointerException{
+        return PathPlanner.loadPath(name, pathConstraints);
+    }
+
+    private PathPoint constructPoint(double x, double y, double rotation, double heading){
+        return new PathPoint(new Translation2d(x, y), Rotation2d.fromDegrees(heading), Rotation2d.fromDegrees(rotation));
+    }
+
+    private PathPlannerTrajectory generateTrajectory(PathPoint firstPoint, PathPoint secondPoint, PathPoint... points){
+        return PathPlanner.generatePath(
+            pathConstraints,
+            firstPoint,
+            secondPoint,
+            points
+        );
+    }
+
+    private Command followTrajectory(String name) throws NullPointerException{
+        //For use with trajectories generated from pathplanner
+        PathPlannerTrajectory trajectory = getTrajectory(name);
+        if(Objects.isNull(trajectory)){throw new NullPointerException();}
+        return followTrajectory(name, trajectory);
+    }
+
+    private Command followTrajectory(String name, PathPlannerTrajectory trajectory){
+        //For use with trajectories generated from a list of poses
+        return new PPSwerveControllerCommand(
+            trajectory,
+            swerveSubsystem::getPose, 
+            DriveConstants.kDriveKinematics, 
+            xController, 
+            yController, 
+            rotationController, 
+            swerveSubsystem::setModuleStates, 
+            swerveSubsystem
+        ).beforeStarting(() -> trajectoryLog.append("Following trajectory " + name)
+        ).andThen(() -> trajectoryLog.append("Trajectory " + name +  " Ended"));
+    }
+
+    private Command resetOdometry(String initialTrajectory) throws NullPointerException{
+        //Resets odometry to the initial position of the given trajectory
+        PathPlannerTrajectory trajectory = getTrajectory(initialTrajectory);
+        Pose2d initialPose = Objects.isNull(trajectory) ? new Pose2d(0, 0, new Rotation2d()) : trajectory.getInitialPose();
+        return new InstantCommand(() -> swerveSubsystem.resetOdometry(initialPose));
```

Comment:
You may want to also reset other odometery sources (e.g. gyro and SwerveModule states). Currently if you try to run a path a second time (by disabling and re-enabling) things go a little haywire